source/fe/mapping_q_cache.cc

```diff
@@ -249,17 +259,465 @@ MappingQCache<dim, spacedim>::initialize(
 {
   AssertDimension(transformation_function.n_components, spacedim);
 
+  this->initialize(mapping,
+                   tria,
+                   [&](const auto &, const auto &point) {
+                     Point<spacedim> new_point;
+                     for (int c = 0; c < spacedim; ++c)
+                       new_point[c] = transformation_function.value(point, c);
+                     return new_point;
+                   },
+                   function_describes_relative_displacement);
+}
+
+
+
+namespace
+{
+  template <typename VectorType,
+            typename std::enable_if<
+              !std::is_same<VectorType,
+                            LinearAlgebra::distributed::Vector<
+                              typename VectorType::value_type>>::value,
+              VectorType>::type * = nullptr>
+  void
+  copy_locally_owned_data_from(
+    const VectorType &vector,
+    LinearAlgebra::distributed::Vector<typename VectorType::value_type>
+      &vector_ghosted)
+  {
+    for (const auto i : vector_ghosted.locally_owned_elements())
+      vector_ghosted[i] = vector[i];
+  }
+
+  template <typename VectorType,
+            typename std::enable_if<
+              std::is_same<VectorType,
+                           LinearAlgebra::distributed::Vector<
+                             typename VectorType::value_type>>::value,
+              VectorType>::type * = nullptr>
+  void
+  copy_locally_owned_data_from(
+    const VectorType &vector,
+    LinearAlgebra::distributed::Vector<typename VectorType::value_type>
+      &vector_ghosted)
+  {
+    vector_ghosted.copy_locally_owned_data_from(vector);
+  }
+
+#if defined(DEAL_II_WITH_TRILINOS) && defined(DEAL_II_WITH_MPI)
+  void
+  copy_locally_owned_data_from(
+    const LinearAlgebra::EpetraWrappers::Vector &vector,
+    LinearAlgebra::distributed::Vector<
+      typename LinearAlgebra::EpetraWrappers::Vector::value_type>
+      &vector_ghosted)
+  {
+    Assert(false, ExcNotImplemented());
+    (void)vector;
+    (void)vector_ghosted;
+  }
+#endif
```

Comment:
Don't we have generic vector access functions? Why doesn't this work with all the standard vector classes?

Reply:
I gave another shot at `LinearAlgebra::ReadWriteVector` and have added two additional methods so that it works for all vectors (see #11938). The code would now look like this:
https://github.com/dealii/dealii/blob/3d64721983a0059b9415d658d0228fab5391fe68/source/fe/mapping_q_cache.cc#L277-L288